### PR TITLE
[SPARK-22256][MESOS] Introduce spark.mesos.driver.memoryOverhead

### PR DESCRIPTION
**Pull request by dmcwhorter**
### What changes were proposed in this pull request?
This is a simple change to support allocating a specified amount of overhead memory for the driver's mesos container.  This is already supported for executors.  

### Why are the changes needed?
This is needed to keep the driver process from exceeding memory limits and being killed off when running on mesos.

### Does this PR introduce _any_ user-facing change?
Yes, it adds a `spark.mesos.driver.memoryOverhead` configuration option.  Documentation changes for this option are included in the PR.

### How was this patch tested?
Test cases covering allocation of driver memory overhead are included in the changes.

### Other notes
This is a second attempt to get this change reviewed, accepted and merged.  The original pull request was closed as stale back in January: https://github.com/apache/spark/pull/21006.
For this pull request, I took the original change by @pmackles, rebased it onto the current master branch, and added a test case that was requested in the original code review.
I'm happy to make any further edits or do anything needed so that this can be included in a future spark release.  I keep having to build custom spark distributions so that we can use spark within our mesos clusters.

### CODE REVIEW

**Review by srowen — COMMENTED** [review]
This looks pretty straightforward. Would it change any behavior if no configuration was set? @skonto @tnachen do you have any opinions on this?

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala [hunk 1]
```diff
@@ -200,6 +201,60 @@ class MesosClusterSchedulerSuite extends SparkFunSuite with LocalSparkContext wi
     })
   }
 
+  test("supports spark.mesos.driver.memoryOverhead with 384mb default") {
```
Comment (HyukjinKwon) [inline resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala:204]:
```suggestion
  test("SPARK-22256: supports spark.mesos.driver.memoryOverhead with 384mb default") {
```

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
@dmcwhorter mind keeping [the GitHub PR template](https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE)?

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala [hunk 2]
```diff
@@ -405,6 +405,20 @@ trait MesosSchedulerUtils extends Logging {
       sc.executorMemory
   }
 
+  /**
+   * Return the amount of memory to allocate to each driver, taking into account
+   * container overheads.
+   *
+   * @param driverDesc used to get driver memory
+   * @return memory requirement as (0.1 * memoryOverhead) or MEMORY_OVERHEAD_MINIMUM
```
Comment (tnachen) [inline resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala:413]: Looks like the actual description here is ->
1. DRIVER_MEMORY_OVERHEAD if set
2. max of MEMORY_OVERHEAD_MINIMUM or 0.1 * driver memory

**Comment by tnachen** [comment]
Besides the comment it LGTM, I think if it's not set it's just using the default config so shouldn't break anything

**Review by srowen — COMMENTED** [review]
@dmcwhorter looking good, just a few comments to respond to it seems

**Comment by dmcwhorter** [comment]
@srowen  I believe if no config is set this change would result in mesos allocating 384MB of extra overhead memory for the driver, whereas previously no extra overhead memory was being allocated.  So I believe this is a behavior change from the previous code.

**Comment by dmcwhorter** [comment]
I just pushed an update to address the comments on the code.  Hopefully this addresses the outstanding concerns.  Please let me know if there are any further changes people would like to see.  Thanks for taking the time to review this.

**Comment by dmcwhorter** [comment]
Also -- I don't believe this change should have introduced the kubernetes integration test failure, but please let me know if that isn't the case and there is something to investigate there.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala [hunk 3]
```diff
@@ -405,6 +405,20 @@ trait MesosSchedulerUtils extends Logging {
...
+   *         (whichever is larger)
```
[the remark this answers is not in the document]
Comment (dmcwhorter) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala]: Yes that's right, updated this comment.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala (hunk 2 above)
Comment (dmcwhorter) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosSchedulerUtils.scala]: Yes, that's right.  Updated this comment.

File: resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala (hunk 1 above)
Comment (dmcwhorter) [reply resource-managers/mesos/src/test/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSuite.scala]: Added this ticket number to both test cases.

**Comment by tnachen** [comment]
@dmcwhorter you're right, we used to just take the memory from the submission. I think it's fine to have this overhead, not sure if we have places to notify important api changes? @srowen

**Comment by srowen** [comment]
It'd be best to put a note in the migration guide for Mesos (if it exists?) or at least `release-notes` label on the JIRA plus a note in the "Docs text" field. It's minor enough as a change/fix that I think it's not out of the question to put it in a minor release. I suppose we can also default the overhead to 0 if that was a concern.

**Comment by dmcwhorter** [comment]
I don't see a migration guide for mesos, unless I'm missing it somewhere.  I also don't think I have access to make those changes to the JIRA.  Could someone else add the release-notes label and a note in the "Docs text" field to the SPARK-22256 ticket?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @dmcwhorter , @srowen , @tnachen , @HyukjinKwon , @viirya .
I added `releaesenotes` label and filled the `Docs Text` for Apache Spark 3.2.0.
Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you for your first contribution, @dmcwhorter . You are added to the Apache Spark contributor group and SPARK-22256 is assigned to you.

**Comment by dmcwhorter** [comment]
Excellent, thanks to all of you for helping me to get this through the process and into spark.

**Pull request closed without merging**